core/src/main/java/io/substrait/relation/HasExtension.java

```diff
@@ -0,0 +1,17 @@
+package io.substrait.relation;
+
+import io.substrait.io.substrait.extension.AdvancedExtension;
+import java.util.Optional;
+
+/**
+ * Used to indicate the presence of an {@link AdvancedExtension} directly on a relation
+ *
+ * <p>This exists outside of {@link Rel} because there are relations that do not have advanced
+ * extensions. See {@link ExtensionLeaf}, {@link ExtensionSingle}, {@link ExtensionMulti}).
+ */
+public interface HasExtension {
+  /**
+   * @return the {@link AdvancedExtension} associated directly with the relation, if present
+   */
+  Optional<AdvancedExtension> getRelExtension();
```

Comment:
how about we move this to general location since other things besides rels can have extensions (and rename to getExtension)

Reply:
To add to this a bit.

The POJO representation effectively merges the contents of `message RelCommon` and the different `message Rel` variants into a single Java object. Effectively we end up with a single object containing two AdvancedExtensions from different sources.

I made the names verbose to help distinguish these, but if you think that `rel.getExtension` and `rel.getCommonExtension` are clear enough I'm happy to change it.

Reply:
I don't have a super strong opinion but it seems like I'd be inclined to have consistency across objects about how to get extensions. For example, it's annoying if it's Plan.getPlanExtension and Rel.getRelExtension as opposed to both of those simply being called getPlanExtension().

In general terms, the way we're using Any is a little weird here. Protobuf has a registry concept. I'm almost more inclined to switch this around and have it be getExtension(Registry) which then returns a list of extensions found as direct objects. This would resolve the double extension pattern we have on Rel/RelCommon and abstract away the use of Any types.

Reply:
Gotcha, so have `getExtension` as the standard accessor, and `getCommonExtension` as a special case for Rels because they collapse two different messages with extension into one class. I like it, I've made the change.

> In general terms, the way we're using Any is a little weird here. Protobuf has a registry concept. I'm almost more inclined to switch this around and have it be getExtension(Registry) which then returns a list of extensions found as direct objects.

I'm less inclined to make this change because I think being able to tell the provenance of the extension is useful. I can imagine a consumer which allows different sets of extensions in RelCommon than on Rels. Even without that, from the user side, it's nice to be able to tell the provenance of the extension from the message, which something like `getExtension(Registry): List<AdvancedExtension>` would throw away.

Additionally, at the relation layer I'm also trying to hide the protobuf layer. I do actually imagine that a common way to implement a custom ProtoRelConverter will be to use the registry mechanism to decode the `any` messages, but by the time we're working with data at the relation layer we (IMO) shouldn't have any protobuf code ideally.